### PR TITLE
Added prop for input attributes - RSC-652

### PR DESCRIPTION
**Pull request by entroform**
https://issues.sonatype.org/browse/RSC-652

### CODE REVIEW

File: gallery/src/components/NxCheckbox/NxCheckboxPage.tsx [hunk 1]
```diff
@@ -94,6 +94,21 @@ const NxCheckboxPage = () =>
...
           </NxTable.Row>
+          <NxTable.Row>
+            <NxTable.Cell>checkboxAttributes</NxTable.Cell>
+            <NxTable.Cell>
+              <NxTextLink external href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox">
+                HTML Input Checkbox Attributes
+              </NxTextLink>
+            </NxTable.Cell>
+            <NxTable.Cell>No</NxTable.Cell>
+            <NxTable.Cell>
```
Comment (rpokorny) [inline gallery/src/components/NxCheckbox/NxCheckboxPage.tsx:105]: I think the API around `checkboxId` is now confusing - why have to pass in the id that way instead of as part of this prop? I'd suggest marking `checkboxId` as deprecated and making it work with either

File: gallery/src/components/NxCheckbox/NxCheckboxPage.tsx [hunk 2]
```diff
@@ -94,6 +94,21 @@ const NxCheckboxPage = () =>
...
+              NxCheckbox also supports any html attribute that's normally supported by the
+              <NxCode>input</NxCode> element except for <NxCode>id</NxCode>, <NxCode>disabled</NxCode>,
```
Comment (rpokorny) [inline gallery/src/components/NxCheckbox/NxCheckboxPage.tsx:107]: The wording here makes it sound like `NxCheckbox` doesn't support externally setting `disabled`.  It does, just not through this prop.  I'd rephrase this description to be less copy/pasted from our usual native attr docs and clearer for the situation at hand.

File: lib/src/components/NxCheckbox/NxCheckbox.tsx [hunk 3]
```diff
@@ -43,15 +45,26 @@ const NxCheckbox = forwardRef<HTMLLabelElement, Props>(
           }),
           content = children && <span className="nx-radio-checkbox__content nx-checkbox__content">{children}</span>;
 
+      const {
+        className: checkboxClassName,
+        ...unfilteredChecboxAttributes
```
Comment (rpokorny) [inline lib/src/components/NxCheckbox/NxCheckbox.tsx:50]:
```suggestion
        ...unfilteredCheckboxAttributes
```

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx [hunk 4]
```diff
@@ -144,4 +144,11 @@ describe('NxCheckbox', function() {
     expect(component).toHaveProp('id', 'foo');
     expect(component).toHaveProp('htmlFor', 'baz');
   });
+
+  it('pass attributes into checkbox input element', function() {
```
Comment (rpokorny) [inline lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx:148]: "it pass attributes into checkbox input element" is not grammatically correct

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx [hunk 5]
```diff
@@ -144,4 +144,11 @@ describe('NxCheckbox', function() {
...
+    expect(getShallowComponent({ checkboxAttributes: { name: 'Garfield'} })
+        .find('input')).toHaveProp('name', 'Garfield');
+    expect(getShallowComponent({ checkboxAttributes: { className: 'cat', name: 'Garfield'} })
+        .find('input')).toHaveClassName('cat');
+  });
```
Comment (rpokorny) [inline lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx:153]: I'd test what happens when something like `disabled` or the id is specified in multiple ways: in the top-level props and also in `checkboxAttributes`.

**Comment by rpokorny** [comment]
I feel like it would be both clearer and more consistent if these new props were all called `inputAttributes` rather than `checkboxAttributes` or `radioAttributes`

File: gallery/src/components/NxCheckbox/NxCheckboxPage.tsx (hunk 1 above)
Comment (entroform) [reply gallery/src/components/NxCheckbox/NxCheckboxPage.tsx:114]: > I think the API around `checkboxId` is now confusing - why have to pass in the id that way instead of as part of this prop? I'd suggest marking `checkboxId` as deprecated and making it work with either

I've updated it in this commit:
4c796beb23b05878976a76dfdcf72610774430aa

File: gallery/src/components/NxCheckbox/NxCheckboxPage.tsx (hunk 2 above)
Comment (entroform) [reply gallery/src/components/NxCheckbox/NxCheckboxPage.tsx]: > The wording here makes it sound like `NxCheckbox` doesn't support externally setting `disabled`. It does, just not through this prop. I'd rephrase this description to be less copy/pasted from our usual native attr docs and clearer for the situation at hand.

Thanks I tried to rephrase it here: 4c796beb23b05878976a76dfdcf72610774430aa

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx (hunk 4 above)
Comment (entroform) [reply lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx]: > "it pass attributes into checkbox input element" is not grammatically correct

Fixed: 4c796beb23b05878976a76dfdcf72610774430aa

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx (hunk 5 above)
Comment (entroform) [reply lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx:169]: > I'd test what happens when something like `disabled` or the id is specified in multiple ways: in the top-level props and also in `checkboxAttributes`.

Updated the tests here: 4c796beb23b05878976a76dfdcf72610774430aa

**Comment by entroform** [comment]
> I feel like it would be both clearer and more consistent if these new props were all called `inputAttributes` rather than `checkboxAttributes` or `radioAttributes`

Thanks! Yeah I thought of using `inputAttributes` as well, but then went with more specific prop name when I saw `checkboxId` and `radioId`.

I updated it to use `inputAttributes` in this commit: 4c796beb23b05878976a76dfdcf72610774430aa

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx [hunk 6]
```diff
@@ -145,10 +146,25 @@ describe('NxCheckbox', function() {
...
+        disabled: false,
+        className: 'checkbox-classname',
+        checked: false
+      }
+    });
+
+    expect(component.find('input')).toHaveProp('id', 'garfield');
+    expect(component.find('input')).toHaveProp('name', 'garfield');
+    expect(component.find('input')).toHaveProp('disabled', true);
+    expect(component.find('input')).toHaveClassName('checkbox-classname');
```
Comment (rpokorny) [inline lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx:167]: Wouldn't hurt to also check that the `input` doesn't have `label-classname`

File: lib/src/components/NxCheckbox/types.ts [hunk 7]
```diff
@@ -4,24 +4,34 @@
...
   checkboxId?: string | null;
   onChange?: (() => void) | null;
   isChecked: boolean;
   disabled?: boolean | null;
   overflowTooltip?: boolean | null;
+  inputAttributes?: InputAttributesProp;
+};
+
+// For testing
+export type PropsWithAnyInputAttributes = Props & {
```
Comment (rpokorny) [inline lib/src/components/NxCheckbox/types.ts:24]: I don't think you need this; you can get away with a cast in the test like this:
```
    const component = getShallowComponent({
      checkboxId: 'not-garfield',
      disabled: true,
      isChecked: true,
      className: 'label-classname',
      inputAttributes: {
        id: 'garfield',
        name: 'garfield',
        disabled: false,
        className: 'checkbox-classname',
        checked: false
      } as Props['inputAttributes']
    });
```

File: lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx (hunk 6 above)
Comment (entroform) [reply lib/src/components/NxCheckbox/__tests__/NxCheckbox.test.tsx]: > Wouldn't hurt to also check that the `input` doesn't have `label-classname`

Agreed! Added test in this commit: d9cf5af435f57153f63578b8eec1a151a520a8cc

File: gallery/src/components/NxCheckbox/NxCheckboxPage.tsx [hunk 8]
```diff
@@ -34,7 +34,16 @@ const NxCheckboxPage = () =>
             <NxTable.Cell>checkboxId</NxTable.Cell>
             <NxTable.Cell>string</NxTable.Cell>
             <NxTable.Cell>No</NxTable.Cell>
-            <NxTable.Cell>An id to identify the checkbox</NxTable.Cell>
+            <NxTable.Cell>
+              An id to identify the checkbox.
+
+              <NxInfoAlert>
```
[the remark this answers is not in the document]
Comment (entroform) [reply gallery/src/components/NxCheckbox/NxCheckboxPage.tsx]: > We usually use warning-level alerts for deprecations

Ah, I was using NxAccordion as an example.
Updated to warning-level alerts here: d9cf5af435f57153f63578b8eec1a151a520a8cc

File: lib/src/components/NxCheckbox/types.ts (hunk 7 above)
Comment (entroform) [reply lib/src/components/NxCheckbox/types.ts]: > I don't think you need this; you can get away with a cast in the test like this:
> 
> ```
>     const component = getShallowComponent({
>       checkboxId: 'not-garfield',
>       disabled: true,
>       isChecked: true,
>       className: 'label-classname',
>       inputAttributes: {
>         id: 'garfield',
>         name: 'garfield',
>         disabled: false,
>         className: 'checkbox-classname',
>         checked: false
>       } as Props['inputAttributes']
>     });
> ```

Fixed: d9cf5af435f57153f63578b8eec1a151a520a8cc

**Review by rpokorny — APPROVED** [review]
